### PR TITLE
[RISCV] Remove experimental from Vector Crypto extensions

### PR DESCRIPTION
**Pull request by ebiggers**
The RISC-V vector crypto extensions have been ratified.  This patch updates the Clang and LLVM support for these extensions to be non-experimental, while leaving the C intrinsics as experimental since the C intrinsics are not yet standardized.

This is an updated version of https://github.com/llvm/llvm-project/pull/69000, which hasn't had activity in a few weeks

### CODE REVIEW

File: clang/include/clang/Support/RISCVVIntrinsicUtils.h [hunk 1]
```diff
@@ -485,7 +485,7 @@ class RVVIntrinsic {
 
 // RVVRequire should be sync'ed with target features, but only
 // required features used in riscv_vector.td.
-enum RVVRequire : uint16_t {
+enum RVVRequire : unsigned int {
```
Comment (topperc) [inline clang/include/clang/Support/RISCVVIntrinsicUtils.h:488]: Use uint32_t since we're counting bits.

File: clang/lib/Basic/Targets/RISCV.h [hunk 2]
```diff
@@ -28,6 +28,7 @@ class RISCVTargetInfo : public TargetInfo {
 protected:
   std::string ABI, CPU;
   std::unique_ptr<llvm::RISCVISAInfo> ISAInfo;
+  bool HasExperimental = false;
```
Comment (topperc) [inline clang/lib/Basic/Targets/RISCV.h:31]: Can this be private like `FastUnalignedAccess`?

Comment (ebiggers) [reply clang/lib/Basic/Targets/RISCV.h]: Done

File: clang/lib/Basic/Targets/RISCV.cpp [hunk 3]
```diff
@@ -370,6 +371,10 @@ bool RISCVTargetInfo::handleTargetFeatures(std::vector<std::string> &Features,
     ISAInfo = std::move(*ParseResult);
   }
 
+  if (std::find(Features.begin(), Features.end(), "+experimental") !=
```
[the remark this answers is not in the document]
Comment (ebiggers) [reply clang/lib/Basic/Targets/RISCV.cpp]: Done using `llvm::is_contained`

File: clang/include/clang/Support/RISCVVIntrinsicUtils.h (hunk 1 above)
Comment (ebiggers) [reply clang/include/clang/Support/RISCVVIntrinsicUtils.h]: Done.  Note, I also changed the type in `RVVIntrinsicRecord::RequiredExtensions` and `SemaRecord::RequiredExtensions`.

**Comment by ebiggers** [comment]
> To be safe, I think we should have an experimental tag to guard the intrinsics, while the extension is ratified and do indeed don't need the experimental.

This is already part of the pull request.

I've updated the commit message and pull request description to make it clear that it's still leaving the C intrinsics as experimental.

> What you propose here seems OK to me, but please add something to RISCVUsage indicating that in some cases an extension may be non-experimental but the intrinsics still experimental and -menable-experimental-extensions must be passed to Clang to use them.

I added a new section "Experimental C Intrinsics" to RISCVUsage.rst.

Also in RISCVUsage.rst, I changed the status of the extensions from "Supported" (the previous proposal) to "Assembly Support".  This is consistent with the fact that the file says that "Supported" means the C intrinsics are supported.

> I also think a Clang release note might make sense, such as "It is now possible for intrinsics to be experimental while the extension itself is non-experimental. This is currently the case for the vector crypto extensions (Zvk*, Zvbb, Zvbc) where -menable-experimental-extension must still be passed to Clang in order to enable use of the C intrinsics."

Done, but I combined the information with the note already being added about the vector crypto extensions, since that seems to be more logical.  Let me know what you think; thanks!

**Comment by ebiggers** [comment]
The `clang-format` warning is from existing code, not from this pull request.

**Review by topperc — APPROVED** [review]
LGTM

**Comment by ebiggers** [comment]
Rebased onto latest upstream because the `clang-format` warning is fixed there.  No changes

File: llvm/docs/RISCVUsage.rst [hunk 4]
```diff
@@ -143,6 +143,22 @@ on support follow.
      ``Zve64f``       Supported
      ``Zve64d``       Supported
      ``Zvfh``         Supported
+     ``Zvbb``         Assembly Support
```
Comment (topperc) [inline llvm/docs/RISCVUsage.rst:146]: Please alphabetize this list correctly.

Comment (ebiggers) [reply llvm/docs/RISCVUsage.rst]: Done

**Review by topperc — APPROVED** [review]
LGTM

**Review by asb — APPROVED** [review]
I was sure I LGTMed this earlier today, but it seems I didn't...

**Comment by ebiggers** [comment]
Rebased to resolve conflicts again

**Comment by topperc** [comment]
@ebiggers are you table to commit or do you need someone else to do it? I don't recall seeing a commit from you before so I wasn't sure.

**Comment by ebiggers** [comment]
No, I don't have commit access.

**Comment by ebiggers** [comment]
Updated to add my Co-authored-by, as the original author is `Brandon Wu <brandon.wu@sifive.com>`.

Someone who has permission to do so will need to approve the GitHub Actions workflow, and merge the commit once everything is ready.

**Comment by ebiggers** [comment]
Thanks @topperc!

**Pull request merged**